### ISSUE TITLE
In the validation images, there are many objects that are not detected

### ISSUE THREAD

**Issue by JosueDavalos** [ultralytics/ultralytics#19790]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

I trained a model to detect berries. I consider the dataset is very representative: around 4k images of hight-resolution with a lot of variability. In general, the object to be detected appears multiple times in each image, averaging around 30-50 instances per image. That is, many instances of different sizes and locations. After training this model, the metrics indicate that the model performs very well. However, when reviewing the test images, I notice that several objects are not detected (it detects about 80% of the objects).

My question is: how can I solve this? Is the model appropriate (YOLOv11)? Should I try another model or adjust some parameters?

btw, I have trained this model (`yolo11s` and `yolo11n`) several times but I still getting the same results

### Environment

```
Ultralytics 8.3.90 🚀 Python-3.12.3 torch-2.5.1 CUDA:0 (NVIDIA GeForce RTX 4070 Ti SUPER, 16376MiB)
Setup complete ✅ (24 CPUs, 95.2 GB RAM, 1804.4/2725.3 GB disk)

OS                  Windows-11-10.0.26100-SP0
Environment         Windows
Python              3.12.3
Install             pip
Path                C:\Users\Josue\miniconda3\envs\gpu\lib\site-packages\ultralytics
RAM                 95.19 GB
Disk                1804.4/2725.3 GB
CPU                 AMD Ryzen 9 7900X3D 12-Core Processor
CPU count           24
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0
{'OS': 'Windows-11-10.0.26100-SP0', 'Environment': 'Windows', 'Python': '3.12.3', 'Install': 'pip', 'Path': 'C:\\Users\\Josue\\miniconda3\\envs\\gpu\\lib\\site-packages\\ultralytics', 'RAM': '95.19 GB', 'Disk': '1804.4/2725.3 GB', 'CPU': 'AMD Ryzen 9 7900X3D 12-Core Processor', 'CPU count': 24, 'GPU': 'NVIDIA GeForce RTX 4070 Ti SUPER, 16376MiB', 'GPU count': 1, 'CUDA': '12.4', 'Package Info': {'numpy': '✅ 1.26.4<=2.1.1,>=1.23.0', 'matplotlib': '✅ 3.9.0>=3.3.0', 'opencv-python': '✅ 4.10.0.84>=4.6.0', 'pillow': '✅ 10.4.0>=7.1.2', 'pyyaml': '✅ 6.0.1>=5.3.1', 'requests': '✅ 2.32.3>=2.23.0', 'scipy': '✅ 1.14.0>=1.4.1', 'torch': '✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.20.1>=0.9.0', 'tqdm': '✅ 4.66.5>=4.64.0', 'psutil': '✅ 5.9.0', 'py-cpuinfo': '✅ 9.0.0', 'pandas': '✅ 2.2.2>=1.1.4', 'seaborn': '✅ 0.13.2>=0.11.0', 'ultralytics-thop': '✅ 2.0.0>=2.0.0'}}
```

### Minimal Reproducible Example

Dataset yaml:
```yaml
path: C:/Users/Josue/strawberry/data/datasets/berry_polygon
train: train
val: test

nc: 1
names:
  0: berry

```

Training code:
```python
from ultralytics.data.augment import Albumentations
from ultralytics.engine.trainer import BaseTrainer
import albumentations as A


def callback_custom_albumentations(trainer: BaseTrainer):
    T = [
        A.OneOf([A.ToGray(), A.RandomGamma(), A.CLAHE()]),
        A.OneOf([A.Blur(blur_limit=(3, 9)), A.MedianBlur(blur_limit=(3, 7))]),
        A.RandomBrightnessContrast(brightness_limit=0.2, contrast_limit=0.2, p=0.25),
        A.HueSaturationValue(
            hue_shift_limit=10, sat_shift_limit=10, val_shift_limit=10, p=0.25
        ),
    ]
    transform = A.Compose(T)

    for t in trainer.train_loader.dataset.transforms.transforms:
        if isinstance(t, Albumentations):
            t.transform = transform
            break

    trainer.train_loader.reset()


dataset_path = "../../data/datasets/berry_dataset.yaml"
wandb.init(project=PROJECT, job_type=RUN_NAME, name=RUN_NAME, notes=DESCRIPTION)
model = YOLO("yolo11s-seg.pt")
model.add_callback("on_pretrain_routine_end", callback_custom_albumentations)

results = model.train(
    data=dataset_path,
    project=PROJECT,
    name=RUN_NAME,
    single_cls=True,
    plots=True,
    batch=32,
    epochs=256,
    imgsz=imgsz,
    seed=7,
    degrees=2,
    dropout=0.2,
    mosaic=0.5,
)

wandb.finish()

```

### Additional

![Image](https://github.com/user-attachments/assets/0010366f-2693-4edf-951f-010da5b46093)

![Image](https://github.com/user-attachments/assets/97d02b54-db61-4f53-b9ba-e66b50235cbb)

![Image](https://github.com/user-attachments/assets/652325d0-bb35-4f1d-b3e8-4400262a1a6d)

![Image](https://github.com/user-attachments/assets/bff385f2-e9f5-4732-b8c4-b53b455cbef6)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JosueDavalos, thank you for your detailed report and for using Ultralytics 🚀!  

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance on training, validation, and troubleshooting. You'll find [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples that might help address your query.  

If this is a 🐛 Bug Report, please ensure that you've provided a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) for us to investigate further. This includes dataset details, code snippets, and relevant logs, which you’ve already shared—thank you!  

For custom training ❓ questions, like this one, please:
- Confirm that your dataset annotations are accurate and consistent.
- Review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), which include optimization strategies for challenging datasets or scenarios with high object density.
- Experiment with other YOLO models (e.g., `yolo11m` or `yolo11l`) to see if a larger model addresses your issue.

Additionally, you can join the vibrant Ultralytics community to get insights and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions.
- [Discourse](https://community.ultralytics.com/) for in-depth conversations.
- [Subreddit](https://reddit.com/r/Ultralytics) for sharing ideas and asking questions.

## Upgrade  

Ensure you’re using the latest version of the `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed. Use the following command in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/):  
```bash
pip install -U ultralytics
```

## Environments  

YOLO runs in various verified environments with all dependencies pre-installed:  
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)  
- **AWS**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)  
- **Docker**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>  

## Status  

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. These tests validate YOLO models across macOS, Windows, and Ubuntu.  

This is an automated response to assist you promptly, but rest assured, an Ultralytics engineer will review your issue and provide additional guidance soon. 😊

**Comment by Y-T-G** [comment]
Only 15 instances are plotted to avoid VRAM issuesm 

https://github.com/ultralytics/ultralytics/issues/19060

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #19792](https://github.com/ultralytics/ultralytics/pull/19792), titled `ultralytics 8.3.94: Limit Segment val plots to 50 items/image for speed`.  

### Key Changes  
This PR introduces significant improvements to validation plotting in YOLO segmentation models:  
- Increased the limit of plotted masks during validation from 15 to 50, enhancing visualization.  
- Added a warning message when more than 50 masks are plotted, helping users understand the performance trade-offs.  
- Standardized the maximum detection limit for plotting predictions to 50 for consistency across functionalities.  

These updates aim to provide better insights into model performance while maintaining efficiency and usability.

### How to Test  
To test if the changes resolve your issue:  
- Use the latest version from PyPI once released.  
- Alternatively, install the updated code directly by running:  
  ```bash
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```  

Your feedback on whether this resolves the issue would be greatly appreciated. Thank you for highlighting this problem, and kudos to the PR author and contributors for their efforts in addressing it! 🙏 If the issue persists, feel free to share additional details or insights.

**Comment by glenn-jocher** [comment]
Hi @JosueDavalos, 

While the visualization issue has been addressed in PR #19792 (increasing the plot limit from 15 to 50 items), your actual detection problem might be related to the challenging nature of your dataset - numerous small objects (30-50 berries) per image.

For improved detection rate, consider:

1. Try larger model variants like YOLO11m or YOLO11l which have better capacity for detecting small objects
2. Adjust your confidence threshold during inference (try lower values like 0.25 or 0.2)
3. Increase your image size (`imgsz`) during both training and inference to capture more detail
4. Experiment with lower IoU thresholds during validation (`--iou 0.5` instead of the default 0.6)
5. Add more augmentations specifically for small object detection like random scaling

For small, numerous objects like berries, the nano and small models might not have sufficient capacity to detect all instances effectively, especially when they overlap or vary greatly in size.
